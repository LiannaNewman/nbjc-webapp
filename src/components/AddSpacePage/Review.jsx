import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import cx from 'classnames';

import Button from '@material-ui/core/Button';
import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';

import StarRating from '../StarRating';

import { addSpaceProps as addSpacePropTypes } from '../../types';

const styles = (theme) => ({
  ratingText: {
    [theme.breakpoints.down('mobile')]: {
      float: 'right',
    },
  },
  review: {
    marginTop: 20,
  },
  remaining: {
    float: 'right',
  },
  checkbox: {
    display: 'block',
    '& p': {
      display: 'inline-block',
    },
  },
  footer: {
    margin: '40px 0',
  },
  submitButton: {
    float: 'right',
    marginBottom: 20,
  },
});

const Review = ({
  classes,
  onBack,
  onNext,
  addSpaceProps,
  title,
  showBack,
  submitLabel,
  disableSubmit,
  overrideClasses,
}) => {
  const matches = useMediaQuery('(min-width:376px)');
  const [formValues, setFormValues] = useState({
    rating: addSpaceProps.rating || 0,
    review: addSpaceProps.review || '',
    anon: addSpaceProps.anon || false,
  });

  const getSanitizedReview = (review) => {
    // todo: sanitize input for invalid characters
    if (review.length > 500) {
      return review.substring(0, 499);
    }
    return review;
  };

  const onChange = (e) => {
    switch (e.target.name) {
      case 'review':
        setFormValues({ ...formValues, review: getSanitizedReview(e.target.value) });
        break;
      case 'anon':
        setFormValues({ ...formValues, anon: !formValues.anon });
        break;
      case 'rating':
        setFormValues({ ...formValues, rating: e.target.value });
        break;
      default:
        break;
    }
  };

  const onFormSubmit = (e) => {
    e.preventDefault();
    onNext(formValues);
  };

  const validateForm = () => formValues.review.length > 0 && formValues.rating > 0;
  // eslint-disable-next-line
  const reviewPlaceholder = 'Did the staff ask for your pronouns? How would you describe the service you received? Did you feel safe entering or being inside the space? Did anyone specific help you? What languages were spoken? Was the space accurate to the listing on The Lavender Book? Was the space accessible? Would you go back?';

  return (
    <>
      <form onSubmit={onFormSubmit} data-testid="review-form">
        {title(matches)}
        <Typography variant="h6">Rate this space</Typography>
        <div>
          <StarRating
            numberFilled={formValues.rating}
            onRatingChanged={(rating) => onChange({ target: { value: rating, name: 'rating' } })}
          />
          <Typography
            variant="caption"
            className={classes.ratingText}
          >
            {`${formValues.rating} stars`}
          </Typography>
        </div>
        <Typography variant="h6">Share your thoughts</Typography>
        <TextField
          value={formValues.review}
          onChange={onChange}
          variant="outlined"
          label="Write a Review"
          helperText={(
            <>
              <Typography variant="caption" color="primary">
                Required
              </Typography>
              <Typography variant="caption" className={classes.remaining}>
                {`${500 - formValues.review.length} characters remaining`}
              </Typography>
            </>
          )}
          placeholder={reviewPlaceholder}
          name="review"
          rows={5}
          className={classes.review}
          data-testid="review-content"
          fullWidth
          multiline
        />
        <FormControlLabel
          className={classes.checkbox}
          checked={formValues.anon}
          control={<Checkbox name="anon" color="primary" onChange={onChange} />}
          label={(
            <Typography variant="body2">
              Publish my rating and review anonymously.
            </Typography>
          )}
        />
        <div className={cx(classes.footer, overrideClasses.footer)}>
          <Button
            type="submit"
            variant="contained"
            color="primary"
            className={cx(classes.submitButton, overrideClasses.submitButton)}
            fullWidth={!matches}
            onClick={onFormSubmit}
            disabled={!validateForm() || disableSubmit()}
            disableElevation
          >
            {submitLabel}
          </Button>
          {showBack && (
            <Button
              type="cancel"
              variant="outlined"
              color="primary"
              fullWidth={!matches}
              onClick={onBack}
              disableElevation
            >
              Back
            </Button>
          )}
        </div>
      </form>
    </>
  );
};

Review.propTypes = {
  classes: PropTypes.shape({}).isRequired,
  onBack: PropTypes.func,
  onNext: PropTypes.func.isRequired,
  addSpaceProps: PropTypes.shape(addSpacePropTypes),
  title: PropTypes.func,
  showBack: PropTypes.bool,
  submitLabel: PropTypes.string,
  disableSubmit: PropTypes.func,
  overrideClasses: PropTypes.shape({}),
};

Review.defaultProps = {
  addSpaceProps: {},
  title: (matches) => (
    <Typography variant={matches ? 'h4' : 'subtitle1'} align="center">
      Rate and review this space. You can rate and review anonymously.
    </Typography>
  ),
  showBack: true,
  onBack: () => {},
  submitLabel: 'Next',
  disableSubmit: () => false,
  overrideClasses: {},
};

export default withStyles(styles)(Review);
