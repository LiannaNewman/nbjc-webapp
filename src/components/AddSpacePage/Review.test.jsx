import React from 'react';
import { render } from '@testing-library/react';

import Review from './Review';

describe('Review', () => {
  test('renders with default values', () => {
    const { asFragment } = render(
      <Review onNext={jest.fn()} onBack={jest.fn()} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
