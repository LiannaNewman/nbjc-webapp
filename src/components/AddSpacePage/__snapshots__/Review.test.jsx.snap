// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review renders with default values 1`] = `
<DocumentFragment>
  <form>
    <h6
      class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-alignCenter"
    >
      Rate and review this space. You can rate and review anonymously.
    </h6>
    <h6
      class="MuiTypography-root MuiTypography-h6"
    >
      Rate this space
    </h6>
    <div>
      <button
        class="MuiButtonBase-root MuiIconButton-root StarRating-button-7 MuiIconButton-colorPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
            />
          </svg>
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root StarRating-button-7 MuiIconButton-colorPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
            />
          </svg>
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root StarRating-button-7 MuiIconButton-colorPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
            />
          </svg>
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root StarRating-button-7 MuiIconButton-colorPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
            />
          </svg>
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root StarRating-button-7 MuiIconButton-colorPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
            />
          </svg>
        </span>
      </button>
      <span
        class="MuiTypography-root Review-ratingText-1 MuiTypography-caption"
      >
        0 stars
      </span>
    </div>
    <h6
      class="MuiTypography-root MuiTypography-h6"
    >
      Share your thought
    </h6>
    <div
      class="MuiFormControl-root MuiTextField-root Review-review-2 MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
      >
        Write a Review
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
          name="review"
          rows="5"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-11"
          >
            <span>
              Write a Review
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained"
      >
        <span
          class="MuiTypography-root MuiTypography-caption MuiTypography-colorPrimary"
        >
          Required
        </span>
        <span
          class="MuiTypography-root Review-remaining-3 MuiTypography-caption"
        >
          500 characters remaining
        </span>
      </p>
    </div>
    <label
      class="MuiFormControlLabel-root Review-checkbox-4"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-13 MuiCheckbox-root MuiCheckbox-colorPrimary MuiIconButton-colorPrimary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-16"
            data-indeterminate="false"
            name="anon"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Publish my rating and review anonymously.
        </p>
      </span>
    </label>
    <div
      class="Review-footer-5"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained Review-submitButton-6 MuiButton-containedSecondary MuiButton-disableElevation Mui-disabled MuiButton-fullWidth Mui-disabled"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Next
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-disableElevation MuiButton-fullWidth"
        tabindex="0"
        type="cancel"
      >
        <span
          class="MuiButton-label"
        >
          Back
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</DocumentFragment>
`;
